### ISSUE TITLE
Loading QMainWindow with QtCompat.loadUi()

### ISSUE THREAD

**Issue by thimic** [mottosso/Qt.py#274]
Hi,

Apologies if this is a stupid question. We have an application loading its main window from a ui file. The code used to look like this:

```python
import sys

from PyQt4 import QtGui, uic

UI_FILE = '/path/to/ui_file.ui'

form_class, base_class = uic.loadUiType(UI_FILE)


class MyApp(base_class, form_class):
    def __init__(self, parent=None):
        super(MyApp, self).__init__(parent=parent)
        self.setupUi(self)

    # Custom code


if __name__ == '__main__':
    app = QtGui.QApplication(sys.argv)
    win = MyApp()
    win.show()
    sys.exit(app.exec_())
```

Ideally, we would like the 'Qt.py' version of this code to look something like this:

```python
import sys

from Qt import QtWidgets, QtCompat

UI_FILE = '/path/to/ui_file.ui'


class MyApp(QtWidgets.QMainWindow):
    def __init__(self, parent=None):
        super(MyApp, self).__init__(parent=parent)
        QtCompat.loadUi(UI_FILE, self)

    # Custom code


if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    win = MyApp()
    win.show()
    sys.exit(app.exec_())
```

This does however raise an exception as `QtWidgets.QMainWindow` gets instantiated with a layout of its own.

I see the example in the docs:

```python
import sys
from Qt import QtCompat

app = QtWidgets.QApplication(sys.argv)
ui = QtCompat.loadUi(uifile="my.ui")
ui.show()
app.exec_()
```

We do however wish to customise our ui class before simply showing it. Is there a way to achieve this without compiling the ui file?

Thanks,
Michael

**Comment by mottosso** [comment]
Hi @thimic, thanks for the question, I'm happy to give this a go, even though Qt Designer isn't my area of expertise.

Is it the case that your UI file already creates a `QMainWindow`, and you wish to customise this as-is, rather than embedding it into another `QMainWindow`?

Would it be an option to customise it directly?

```python
...
ui = QtCompat.loadUi(uifile="my.ui")
ui.setWindowTitle("My custom title")
ui.show()
```

You can add and edit widgets that way, by simply pretending that `ui` is the `self` you would normally have access to in the `__init__` method. 

Otherwise, I think what I would try, is not creating your QMainWindow from Qt Designer, but rather create something you can embed into it, like a main widget. That way, you can do as you have done, and then embed it once loaded.

Not sure if that's helpful. :)

**Comment by MHendricks** [comment]
The way you want to do it should work. That's how I normally use loadUi.

> This does however raise an exception as QtWidgets.QMainWindow gets instantiated with a layout of its own.

Are you using the latest version? Before v1.1.0.b9 QtCompat.loadUi would incorrectly error out because QMainWindow and QDockWidgets automatically create a layout on creation. See PR #217 for more info on what was changed.

**Comment by thimic** [comment]
Thanks for the quick replies :) 

@MHendricks, That should solve it! We're currently on 1.1.0.b6. I should have checked that, my apologies. I looked through the PRs, but must have missed #217. I'll make sure to run the most recent version from now on.

@mottosso, Your assumption is correct. We are re-implementing our own methods on the QMainWindow, so referring to the instantiated `ui` object isn't quite enough in our case. Though looks like it's already been fixed in #217. 

Thanks a lot for your work. Qt.py is pretty awesome! :)
